sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -309,20 +309,41 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
    * UnsafeRow is highly compressible (at least 8 bytes for any column), the byte array is also
    * compressed.
    */
-  private def getByteArrayRdd(n: Int = -1): RDD[(Long, Array[Byte])] = {
+  private def getByteArrayRdd(n: Int = -1, reverse: Boolean = false): RDD[(Long, Array[Byte])] = {
     execute().mapPartitionsInternal { iter =>
       var count = 0
       val buffer = new Array[Byte](4 << 10)  // 4K
       val codec = CompressionCodec.createCodec(SparkEnv.get.conf)
       val bos = new ByteArrayOutputStream()
       val out = new DataOutputStream(codec.compressedOutputStream(bos))
-      // `iter.hasNext` may produce one row and buffer it, we should only call it when the limit is
-      // not hit.
-      while ((n < 0 || count < n) && iter.hasNext) {
-        val row = iter.next().asInstanceOf[UnsafeRow]
-        out.writeInt(row.getSizeInBytes)
-        row.writeToStream(out, buffer)
-        count += 1
+
+      if (reverse) {
+        // To collect n from the last, we should anyway read everything with keeping the n.
+        // Otherwise, we don't know where is the last from the iterator.
+        var last: Seq[UnsafeRow] = Seq.empty[UnsafeRow]
+        if (n > 0) {
+          // NOTE: when reverse, -1 is not supported. There is no difference with collecting
+          // all and reversing.
+          val slidingIter = iter.map(_.copy()).sliding(n)
```

Comment:
Using `sliding` does make the code much simpler, but it also makes it hard to understand the performance here.

IIUC, `Iterator.sliding` uses `ArrayBuffer` to buffer the elements, and keeps removing elements from the beginning, adding elements at the end. This triggers array copy every time.

I think it's better to implement this logic manually and use a FIFO queue to buffer the results.

Reply:
I tired it with a finite queue and couldn't find a significant performance gain (https://github.com/apache/spark/pull/26809#discussion_r355370827).

Reply:
And, actually I think it's best for Scala itself to fix it if that matters.

Reply:
makes sense